### PR TITLE
Adaptive dmd

### PR DESCRIPTION
**Pull request by kevinhkhuynh**
AdaptiveDMD, a version of DMD that works around varying dt, by using interpolation to interpolate snapshots to make them constant dt.

### CODE REVIEW

**Comment by dylan-copeland** [comment]
Is there an example that uses this new feature?

**Review by siuwuncheung — APPROVED** [review]

File: lib/algo/AdaptiveDMD.C [hunk 1]
```diff
@@ -0,0 +1,124 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2013-2021, Lawrence Livermore National Security, LLC
```
Comment (dylan-copeland) [inline lib/algo/AdaptiveDMD.C:3]: Should the year be updated to 2022 in another PR, to keep this one simple?

File: lib/algo/AdaptiveDMD.h [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+// Description: Computes the AdaptiveDMD algorithm on the given snapshot matrix.
+//              The AdaptiveDMD algorithm should be used if the dt changes
+//              between the samples. This algorithm interpolates the ununiformly
+//              sampled samples such that they are uniformly distanced and the
```
Comment (dylan-copeland) [inline lib/algo/AdaptiveDMD.h:14]: This could be written more clearly, for example: "This algorithm uniformly interpolates the samples that may have been taken with variable steps, using the constant step dt."

File: lib/algo/AdaptiveDMD.h [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    const Matrix* interpolateSnapshots();
+
+    /**
+     * @brief The stored times of each sample.
+     */
+    std::vector<Vector*> d_sampled_times;
+
+    /**
+     * @brief The dt to uniformly interpolate the samples between.
```
Comment (dylan-copeland) [inline lib/algo/AdaptiveDMD.h:100]: Better wording: "The constant step size for uniform interpolation of samples."

File: lib/algo/manifold_interp/Interpolator.h [hunk 4]
```diff
@@ -150,6 +127,39 @@ class Interpolator
...
+ */
+std::vector<double> obtainRBFToTrainingPoints(std::vector<Vector*> parameter_points,
+    std::string interp_method, std::string rbf, double epsilon, Vector* point);
+
+/**
+ * @brief Compute the sum of the RBF weights.
+ *
+ * @param[in] rbf The vector holding the rbfs of the training points.
+ */
+double rbfWeightedSum(std::vector<double> rbf);
```
Comment (dylan-copeland) [inline lib/algo/manifold_interp/Interpolator.h:150]: Here and in some other places, `rbf` is passed by value. Is it certain to always be a small `vector`?

File: lib/algo/manifold_interp/VectorInterpolator.C [hunk 5]
```diff
@@ -74,117 +74,140 @@ void VectorInterpolator::obtainLambda()
...
+    {
+
+        for (int i = 0; i < d_parameter_points.size(); i++)
         {
-            for (int j = 0; j < f_T->numColumns(); j++)
+            // For ref point, gamma is the zero vector.
+            if (i == d_ref_point)
             {
-                f_T->item(i, j) = d_gammas[j]->getData()[i];
+                Vector* gamma = new Vector(d_rotated_reduced_vectors[d_ref_point]->dim(), d_rotated_reduced_vectors[d_ref_point]->distributed());
```
Comment (dylan-copeland) [inline lib/algo/manifold_interp/VectorInterpolator.C:96]: long line

File: lib/algo/manifold_interp/VectorInterpolator.C [hunk 6]
```diff
@@ -74,117 +74,140 @@ void VectorInterpolator::obtainLambda()
...
-        d_lambda_T = f_T;
-    }
+    // The exp mapping is X + the interpolated gamma
+    Vector* interpolated_vector = d_rotated_reduced_vectors[d_ref_point]->plus(log_interpolated_vector);
+    delete log_interpolated_vector;
+    return interpolated_vector;
 }
 
-Vector* VectorInterpolator::obtainLogInterpolatedVector(std::vector<double> rbf)
+Vector* obtainInterpolatedVector(std::vector<Vector*> parameter_points, std::vector<Vector*> data, Matrix* f_T, std::string interp_method, std::vector<double> rbf)
```
Comment (dylan-copeland) [inline lib/algo/manifold_interp/VectorInterpolator.C:123]: long line

File: lib/algo/manifold_interp/VectorInterpolator.C [hunk 7]
```diff
@@ -74,117 +74,140 @@ void VectorInterpolator::obtainLambda()
...
             }
         }
     }
 
-    return log_interpolated_vector;
+    return interpolated_vector;
 }
 
-Vector* VectorInterpolator::interpolate(Vector* point)
+Matrix* solveLinearSystem(std::vector<Vector*> parameter_points, std::vector<Vector*> data, std::string interp_method, std::string rbf, double& epsilon)
```
Comment (dylan-copeland) [inline lib/algo/manifold_interp/VectorInterpolator.C:162]: long line

File: lib/algo/manifold_interp/VectorInterpolator.h [hunk 8]
```diff
@@ -112,6 +112,10 @@ class VectorInterpolator : public Interpolator
     std::vector<Vector*> d_gammas;
 };
 
+Vector* obtainInterpolatedVector(std::vector<Vector*> parameter_points, std::vector<Vector*> data, Matrix* f_T, std::string interp_method, std::vector<double> rbf);
```
Comment (dylan-copeland) [inline lib/algo/manifold_interp/VectorInterpolator.h:115]: Long lines

**Review by dylan-copeland — CHANGES_REQUESTED** [review]

**Comment by kevinhkhuynh** [comment]
> Is there an example that uses this new feature?

will add for ex 18

**Comment by kevinhkhuynh** [comment]
Need to add comments about LS, IDW, LP, and epsilon

**Comment by kevinhkhuynh** [comment]
> Is there an example that uses this new feature?

AdaptiveDMD is now used in dg_euler. I guess regular DMD should never have been used there in the first place, since the dt is changing.

File: lib/algo/AdaptiveDMD.C (hunk 1 above)
Comment (kevinhkhuynh) [reply lib/algo/AdaptiveDMD.C:3]: I think that would be better.

**Comment by kevinhkhuynh** [comment]
RFR again.

File: lib/algo/manifold_interp/VectorInterpolator.C [hunk 9]
```diff
@@ -1,6 +1,5 @@
 /******************************************************************************
  *
- * Copyright (c) 2013-2021, Lawrence Livermore National Security, LLC
```
Comment (dylan-copeland) [inline lib/algo/manifold_interp/VectorInterpolator.C:3]: The first line is missing.

**Comment by dylan-copeland** [comment]
The two files `lib/hyperreduction/STSampling.*` still need 2019 -> 2021.

**Review by dylan-copeland — APPROVED** [review]
